spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Normalize.scala

```diff
@@ -69,13 +69,49 @@ class Normalize[T: ClassTag](val p: Double, val eps: Double = 1e-10
       } else {
         buffer.resizeAs(inputBuffer).pow(inputBuffer, ev.fromType(p))
       }
-      normp.sum(buffer, 2).add(ev.fromType(eps))
+      // normp.sum(buffer, 2).add(ev.fromType(eps))
+      // perf fix: the sum operation of tensor will call Java + element wise
+      // perf fix: start
+      normp.resize(Array(buffer.size(1), 1, buffer.size(3), buffer.size(4)))
+      var batchSize = 0
+      while (batchSize < normp.size(1)) {
+        val batchOfNormp = normp.select(1, batchSize + 1).zero
+        val batchOfInput = buffer.narrow(1, batchSize + 1, 1)
+
+        var channel = 0
+        while (channel < buffer.size(2)) {
+          batchOfNormp.add(batchOfInput.select(2, channel + 1))
+          channel += 1
+        }
+        batchSize += 1
+      }
+      normp.add(ev.fromType(eps))
+      // perf fix: end
       norm.resizeAs(normp).pow(normp, ev.fromType(1.0 / p))
     }
     if (norm.dim() <= 2) {
       output.cdiv(inputBuffer, norm.view(norm.nElement(), 1).expandAs(inputBuffer))
     } else if (norm.dim() == 4) {
-      output.cdiv(inputBuffer, norm.view(norm.size()).expandAs(inputBuffer))
+      // output.cdiv(inputBuffer, norm.view(norm.size()).expandAs(inputBuffer))
```

Comment:
remove it ?

Reply:
I don't think we should remove it. Because this description is much simpler to understand how does this layer do. But the reason I did this optimization because of the `norm` will be expanded as an uncontiguous Tensor. So I had to compute it channel by channel.

Reply:
Which one is uncontiguous ? If just for contiguous, you can convert uncontiguous to contiguous.

Reply:
@zhangxiaoli73 I have tested, the conversion will take too much time and consume a lot of memory.